chore(workflows): upgrade actions/cache from v2 to v4

Aire-One opened the pull request:
actions/cache@v2 is being deprecated, and older versions are removed: https://github.com/actions/cache?tab=readme-ov-file#%EF%B8%8F-important-changes

Only updating the version seems to be enough to unblock the CI (running workflows on my fork was a success https://github.com/Aire-One/awesome/pull/3)

sclu1034 commented:
Given [recent events](https://www.stepsecurity.io/blog/harden-runner-detection-tj-actions-changed-files-action-is-compromised), actions should better be pinned to SHAs now, not tags.

And #3994 will need to be fixed until April as well.

Aire-One commented:
```
There were 1 errors:
 - /home/runner/work/awesome/awesome/tests/test-awful-layout.lua: 2025-03-29 02:32:38 E: awesome: Error during a protected call: lib/awful/layout/init.lua:267: bad argument #-2 to 'geometry' (value in [-32768.0, 32767.0] expected, got 65576.0)
 ```

Well, this one is going to be funny...

Aire-One commented:
Sooooo it took me some additional researching and try-and-fails, but it actually seems like there was an issue with the memory management from LuaJIT with lightuserdata. Using an updated version of LuaJIT makes the error vanish.

Also, updating the Actions for CodeCov unblocked it! I'm making big progress :rocket:

Aire-One commented:
Hello, I've rebased/pushed the working commits from my private branch to here. The CI is working again!

Think that should be noted : 

- As mentioned earlier here, and described in the commit message for 73f551a26b9fc321fbba5272841536c7bb1fb16b: This is just a mitigation patch to allow the test to pass. I couldn't figure out the root cause for this error.
- The `Update API docs` workflow is green, but the script actually fails. The error is that the secret token used is only available in the main repository. When submitting a PR, the workflow run as from your fork, hence don't have access to the token. I was induced in error thinking something was wrong because lasts PRs from Elv had some comments from the "awesome bot". This is because they were branches from this repo and not his fork. So ultimately, we'll see if the script works only after merge, with the workflow triggered from master.
- I used the `luarocks/gh-actions-lua` to install LuaJIT from source in the CI. LuaJIT uses rolling release, so it sounds best suited to always (and only) test against the git-master version. If really needed, we can add (with new PRs) entries in the workflow matrix to test against legacies versions. In the long term, I think it would be benefited for us to use this Actions to install all Lua versions (and also to use the other Actions to install Luarocks).
- There is a new "Correct Lua variables paths" step in the main workflow, it is needed to use the `github.workspace` variable. I corrected both `LUAINCLUDE` and `LUALIBRARY`, but only the fix for `LUAINCLUDE` is strictly necessary. The path fix is needed to build Luarocks. I have kept the `--with-lua-include` flag, but I'm pretty sure we actually don't need it anymore with the `gh-actions-lua`.
- We may have issue with how CodeCov calculate the coverage. It was unnoticed with https://github.com/awesomeWM/awesome/pull/3947 that says `- Coverage   master:91.26%   diff:48.52%   +/-:-42.74%`. So we have somehow lost about half of the coverage.
- With the very last debug session, I had a new error with integration tests with Lua5.2 `/home/runner/work/awesome/awesome/tests/test-awful-client.lua: Error: timeout waiting for signal in step 11/138 (@20) (.../runner/work/awesome/awesome/tests/test-awful-client.lua:221).` If it doesn't happen here, I'll skip debugging it now.

actionless reviewed (approved):
👌😸🔥

Aire-One commented:
Thank you for the merge @actionless :rocket:

actionless commented:
considering that @Elv13 previously approved the prev version of the PR i think there is no need to futher block the CI, so finally we could merge some of long-waiting PRs

thanks again for your great pipeline investigation, @Aire-One 🚀
